src/main/java/org/craftercms/studio/impl/v1/service/webdav/WebDavServiceImpl.java

```
@@ -180,22 +180,32 @@ public String upload(@ValidateStringParam(name = "site_id") final String site,
                          final InputStream content)
         throws WebDavException {
         WebDavProfile profile = profileReader.getProfile(site, profileId);
-        String uploadUrl = profile.getBaseUrl();
+        String uploadUrl = StringUtils.removeEnd(profile.getBaseUrl(), "/");
         try {
+            Sardine sardine = SardineFactory.begin(profile.getUsername(), profile.getPassword());
+
             if(StringUtils.isNotEmpty(path)) {
-                uploadUrl +=  path.startsWith("/")? path : "/" + path;
+                String[] folders = StringUtils.split(path, "/");
+
+                for(String folder : folders) {
+                    uploadUrl += "/" + folder;
+
+                    try {
+                        logger.debug("Trying to create folder {0}", uploadUrl);
+                        sardine.createDirectory(uploadUrl);
+                        logger.debug("Folder created");
+                    } catch (Exception e) {
+                        logger.debug("Folder already exists");
```

Wouldn't it be better to first check if the dir exists and then create it? Instead of trying to create it and then ignore the exception if it already exists. Also what if the exception is a connection issue?

---

This was implemented in this way because webdav only allows checking if files exist but not folders, so the only way was to try to create it and if an exception is thrown it means it already exists, about connectivity issues the current library doesn't provide specific exceptions to distinguish each case.

---

Understood. Maybe we can log the exception too and put a debug message of `Folder already exists or it couldn't be created`?

---

Actually I think there is a way to distinguish them, I didn't notice because its a runtime exception, I will improve this

---

Ok good

---

Also when  logging add the `uploadUrl` to the message. Eg. `"Folder " + uploadUrl + " created"`